### ISSUE TITLE
DRAM-setup.py with sqlite3.IntegrityError

### ISSUE THREAD

**Issue by dot4822** [shafferm/DRAM#58]
Hi,

I have installed DRAM via Bioconda using command line `conda create -n DRAM -c conda-forge -c bioconda dram`.

After installation, I am trying to set up DRAM via `DRAM-setup.py prepare_databases --output_dir DRAM_data --skip_uniref --threads 20`.

However, I have got an error during processing viral_description.id. The error message is shown as below:

`sqlite3.IntegrityError: UNIQUE constraint failed: viral_description.id
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: viral_description.id
[SQL: INSERT INTO viral_description (id, description) VALUES (?, ?)]
[parameters: (('YP_009966115.1', 'YP_009966115.1 homing endonuclease [Escherichia virus H8]'), ('YP_009966083.1', 'YP_009966083.1 phage tail protein [Escherichia virus H8]'), ('YP_009966051.1', 'YP_009966051.1 exonuclease (ACLAME 70) [Escherichia virus H8]'), ('YP_009966019.1', 'YP_009966019.1 transmembrane protein [Escherichia virus H8]'), ('YP_009965987.1', 'YP_009965987.1 hypothetical protein FGG59_gp021 [Escherichia virus H8]'), ('YP_009952253.1', 'YP_009952253.1 major capsid protein [Mycobacterium phage Adephagia]'), ('YP_009952285.1', 'YP_009952285.1 hypothetical protein I5G89_gp55 [Mycobacterium phage Adephagia]'), ('YP_009952317.1', 'YP_009952317.1 hypothetical protein I5G89_gp23 [Mycobacterium phage Adephagia]')  ... displaying 10 of 6660 total bound parameter sets ...  ('PH', 'PH'), ('YMVPNPNGSTGIFVVVQVRLMHPRYSIAMGF9HNGQSLGVHVIGGDGVINMEVSEGEAGSELTICVLSTQGIDCEHCGGHSDLNPQSSnNSAGHSNWQDLSMDTPTADGIAFSP', 'YMVPNPNGSTGIFVVVQVRLMHPRYSIAMGF9HNGQSLGVHVIGGDGVINMEVSEGEAGSELTICVLSTQGIDCEHCGGHSDLNPQSSnNSAGHSNWQDLSMDTPTADGIAFSP'))]
(Background on this error at: http://sqlalche.me/e/13/gkpj)`

I wonder whether anyone has met the same error, and how to deal with this issue.

Thanks!

Regards,
Dot

**Comment by mw55309** [comment]
Try the parse_viral.pl script that I've put here: https://github.com/mw55309/DRAM_hacks

Then check to see if any of the IDs are duplicated?

**Comment by dot4822** [comment]
Hi Mick,

Thanks for the reply.

I have tried the perl script as you suggested. It went well at first (e.g. ```"YP_009481243.1","YP_009481243.1 hypothetical protein pmac_cds_559 [Pandoravirus macleodensis]"```).

But then the output is not readable, and is shown as below:

```
"AQIAALHGDNICGMMDWCCCYSMVTMPAMDCALAACVACFHCATCF","AQIAALHGDNICGMMDWCCCYSMVTMPAMDCALAACVACFHCATCF"
"YGFRSENGNADng","YGFRSENGNADng prFKNCRRWWPLCSSHGGACVVTVALCFEg pDDTGGNGDA"
"YypotFARRY","YypotFARRY"
"AQIAALHGDNICGMMDWCCCYSMVTMPAMDCALAACVACFHCATCF","AQIAALHGDNICGMMDWCCCYSMVTMPAMDCALAACVACFHCATCF"
"YQQFLRLandoHRH","YQQFLRLandoHRH"
"YTDGC","YTDGC"
"PH","PH"
"GLFTGKRTYGVLMRetiSWGAEDAYSPKHTVIBBaESSVSFAYRKAITYLD","GLFTGKRTYGVLMRetiSWGAEDAYSPKHTVIBBaESSVSFAYRKAITYLD"
"EELPDIIFECSuGEEPERRDIAYVVVGIAEHFVMQMHFGYoAKAREAGGCHPSNNALETPLVEVAACVVme","EELPDIIFECSuGEEPERRDIAYVVVGIAEHFVMQMHFGYoAKAREAGGCHPSNNALETPLVEVAACVVme ATELRYPGHMAensIIVIVQRSIQNhypVQQNHITASuGETAFASIM>YP_VAAEPQMPl pr"
"AQIAALHGDNICGMMDWCCCYSMHGLLITFTA","AQIAALHGDNICGMMDWCCCYSMHGLLITFTA"
"ILEVGSKRALTHMAeTFKTGLHVPRAKDWLASKTAc_5LNDWTTFFQLLTASHQQIGDWNL094TRQWILVEYESlWNHGRVLcalDLIGTILDAPDNILDNTERVWTLLERPAVDRRVLAAIGPYDHl","ILEVGSKRALTHMAeTFKTGLHVPRAKDWLASKTAc_5LNDWTTFFQLLTASHQQIGDWNL094TRQWILVEYESlWNHGRVLcalDLIGTILDAPDNILDNTERVWTLLERPAVDRRVLAAIGPYDHl SYP_VRYP>YP_AVVVSERIWQQFPNGSDTLTMPNHKWCMHPNIQGDNIVILARWVTNED"
"YP_YdorPLFKTERDMMRRRRFVRPHHINYHADtAERLHAGTDKPNHPIVPaiNRSYCQDHAAHNMCLIFIDQVLAALVGLEYLTAFYVDCSDPKSELAAAFPALADLPSDDTCDDACEAAAP","YP_YdorPLFKTERDMMRRRRFVRPHHINYHADtAERLHAGTDKPNHPIVPaiNRSYCQDHAAHNMCLIFIDQVLAALVGLEYLTAFYVDCSDPKSELAAAFPALADLPSDDTCDDACEAAAP"
"GQSuGSYAAAHIAPVTTMCIQSIeiGVARTYPDDRTFKVLodNMSG4812T948GTDGRSVLCVNGGHAToPSRQYEAMCN","GQSuGSYAAAHIAPVTTMCIQSIeiGVARTYPDDRTFKVLodNMSG4812T948GTDGRSVLCVNGGHAToPSRQYEAMCN"
"DGREPHQSIDL094NI>AQIVDRCPRYHCTclDDACMLHAFPVPaiNP7.LTRLLPGRESPVTDN","DGREPHQSIDL094NI>AQIVDRCPRYHCTclDDACMLHAFPVPaiNP7.LTRLLPGRESPVTDN"
"AQIAALHGDNICGMMDWIVEMVPWVEToPSLCFDILKHKMRSHPALMHIAADWTSRIHNNHQIPSPLQVILGRRVSDPKLDNVTRNNPRRRRRATLFFQLNGyKSAWEMRMGacQFDAVKNLPRCPTEGRGDLHDESDEYSiILDAAKAGF","AQIAALHGDNICGMMDWIVEMVPWVEToPSLCFDILKHKMRSHPALMHIAADWTSRIHNNHQIPSPLQVILGRRVSDPKLDNVTRNNPRRRRRATLFFQLNGyKSAWEMRMGacQFDAVKNLPRCPTEGRGDLHDESDEYSiILDAAKAGF inADtIRDCGLmaPGSSGMSNHHE9TRNCTQRVIILKHMISVETASAVEPA"
"EVCYMQDkFGPQWK0GdeFRFKAPRQPAEAKCRHAVNRTGG.994DENIYANHGRHPNQGGRNSVLRCVVLNSYELK>GLFCLRRRMFMDFPTAEALLMaEAEQIIGDRIDYVYGMYLSTNAPleVKAALHIACVGCTHALDGHDRFSNEEIACDCSNPYRATMrEWATKLCALDYVYDYVYDSRYLSVYin","EVCYMQDkFGPQWK0GdeFRFKAPRQPAEAKCRHAVNRTGG.994DENIYANHGRHPNQGGRNSVLRCVVLNSYELK>GLFCLRRRMFMDFPTAEALLMaEAEQIIGDRIDYVYGMYLSTNAPleVKAALHIACVGCTHALDGHDRFSNEEIACDCSNPYRATMrEWATKLCALDYVYDYVYDSRYLSVYin PPICANWRHAESDSVARKTHGPWGQKGCPMN2CSEVFCNEQyyDTDAPNGSAWprGSdomaWLPHIDLESEHCDSDSDSSAAIDTRNCDVPAPRMLAALNIS"
"YVHHDGRSDtaiK","YVHHDGRSDtaiK"
"APARRCLAnVAMERGQADPPNPTGYGPGHQGLRVIIGLPFDTGYVWcdsGQLPFDSNEYE2NIG","APARRCLAnVAMERGQADPPNPTGYGPGHQGLRVIIGLPFDTGYVWcdsGQLPFDSNEYE2NIG"
"YPDLCHleSRRKGCPCSKPGIGaclEADPRRIEKMYDLAR","YPDLCHleSRRKGCPCSKPGIGaclEADPRRIEKMYDLAR"
"LDVN","LDVN"
"GPTDI","GPTDI"
"MRAIVDYGVTPLITFSAFCGGVKGYESAGVSLHFINYCTALHGGLA>YP_DLSGLELIYKGPTAAhILFVDFLGALRLPDAGT","MRAIVDYGVTPLITFSAFCGGVKGYESAGVSLHFINYCTALHGGLA>YP_DLSGLELIYKGPTAAhILFVDFLGALRLPDAGT"
"AGWLHHTERNIVCK","AGWLHHTERNIVCK"
"GQF7TGGHAHRPRELSLCNM8GTDLWDAIVNWWGES","GQF7TGGHAHRPRELSLCNM8GTDLWDAIVNWWGES"
"YKTRCVLLRMLTAFPMLHVGALRQWDGNHAPQLGEDVPPSTLL","YKTRCVLLRMLTAFPMLHVGALRQWDGNHAPQLGEDVPPSTLL"
"YKRTDNALDPYMKRrPANASH","YKRTDNALDPYMKRrPANASH"
"TIGNTHAHECTT>GQFGAPQLVYYDTVKHGDMI","TIGNTHAHECTT>GQFGAPQLVYYDTVKHGDMI"
"TEPERLTGHSHWHERLMLSCQHERLVVrGDKIMGEK8DGEAnDIEVVIRGDQNPEA","TEPERLTGHSHWHERLMLSCQHERLVVrGDKIMGEK8DGEAnDIEVVIRGDQNPEA"
"GLDEEIPHQALPNGLAMNEDLQWKYIGW09GERAYRCPLT","GLDEEIPHQALPNGLAMNEDLQWKYIGW09GERAYRCPLT"
"YKRTDNALDPYMSTVLodenSRDNSLWMGGVEMNAGSPDDRECSECK","YKRTDNALDPYMSTVLodenSRDNSLWMGGVEMNAGSPDDRECSECK"
"LMLCDDDNrHVANDNSFHSAEAnLTNVTNSFRDLSLRMLCKSGRYKRGLLPPSDTThhhhQyyyDMRLHNFPNMYHPWDVPFTPIVMIHGNHAMEs_EACGSEHCRTYKGPTEDLQ8QRRFMSLDPECLSMGLA>TKCMHEYtFIGPAWNGHlYVNLEHYQGPECMeo","LMLCDDDNrHVANDNSFHSAEAnLTNVTNSFRDLSLRMLCKSGRYKRGLLPPSDTThhhhQyyyDMRLHNFPNMYHPWDVPFTPIVMIHGNHAMEs_EACGSEHCRTYKGPTEDLQ8QRRFMSLDPECLSMGLA>TKCMHEYtFIGPAWNGHlYVNLEHYQGPECMeo"
"QKDFEHDVYLWPQSTAVWGEK8","QKDFEHDVYLWPQSTAVWGEK8"
"LCTLEyWGGACFIKCGRVSTR>QKDSVRAFNH4FGVVWRVWARQCGGHVHC","LCTLEyWGGACFIKCGRVSTR>QKDSVRAFNH4FGVVWRVWARQCGGHVHC"
"AF>YVHDLEIY81HHPAGPDDARVVTVGTDNVPITGSH","AF>YVHDLEIY81HHPAGPDDARVVTVGTDNVPITGSH"
"YP_009481243.1","YP_009481243.1 hypothetical protein pmac_cds_uIGTuHCRVVKQYEFERRDPTLTEEELRFITSIFQDPAFLVACERQYELVDKVPGLIHLFHTQMVALNNyANFGHDQGFANYTGSHRLRI"
"RFEVAEQIHHEKAIandNGMYCHMLRPDPTAFERRPYKIIQFtFRWGCMWSRGRDCDAMCNGTHIIFYWIVSTLVPLFKSVSLGTGVSSGVHCVVQLDA>QKDEIQIFRTTVPVATLDLEHIKQVQLDEGFG","RFEVAEQIHHEKAIandNGMYCHMLRPDPTAFERRPYKIIQFtFRWGCMWSRGRDCDAMCNGTHIIFYWIVSTLVPLFKSVSLGTGVSSGVHCVVQLDA>QKDEIQIFRTTVPVATLDLEHIKQVQLDEGFG [PaILHGVQWTRRandorHVALMTyHPSI47LFLHLFHMRhhhDLPPTCKEIVEHALVLSPVVP TEAPHDLKYVAQIAVYFPIAGQTYFPIAa7DWdWRPHVDWGELLRSSGFGQa8HPKRAGGLNSTVPLFFILPTKRMISVEYVAQENMLMDAAICMYIEASMYIALMAPLCTLTWEPTYTGVSIPE.SKDLIGRCaH2AVQNVLPGWVAYMQQICMI TEAPPPPPMEGANERVWAPYILRTQAMVLYGISVRVprtSWMGQTISRGVVSERLPALMSGYQPFYDTitoravGMIWLSSDDSPHRyyNenQTTPLTWWIpM"
"GYELI5YQPVPLFFILPTKRMISVEYVAQENMLMDAAICMYIE2IPRTCAPLACFIKIQCFIRMSYARleWDEHIRGPV1GSDQGHNGDCGKNETLWATAPNNEDMASRAGMEGADRNNQEVAMTCRVC21CSC26HPSQNMSILNPACSYELNDADIIQFCG48FIV","GYELI5YQPVPLFFILPTKRMISVEYVAQENMLMDAAICMYIE2IPRTCAPLACFIKIQCFIRMSYARleWDEHIRGPV1GSDQGHNGDCGKNETLWATAPNNEDMASRAGMEGADRNNQEVAMTCRVC21CSC26HPSQNMSILNPACSYELNDADIIQFCG48FIV"
"SDLLATEgNLa","SDLLATEgNLa"
"MLWWAEIQPGAPDN>YP_ELNDVYFLLVS","MLWWAEIQPGAPDN>YP_ELNDVYFLLVS"
"YP_00948124QNHVAFc_cPVKHKKaMSAHDSI","YP_00948124QNHVAFc_cPVKHKKaMSAHDSI"
"VK4QEKIEIEWRQLQNQMRWLISVEPKNETHELKVKIQWLPLNEWCMQLSAYrSIIMEQMSQNL>YVHWGyKDGacCCdHYQWATAIL","VK4QEKIEIEWRQLQNQMRWLISVEPKNETHELKVKIQWLPLNEWCMQLSAYrSIIMEQMSQNL>YVHWGyKDGacCCdHYQWATAIL"
"AEA>YPDSW","AEA>YPDSW"
"AF>YVHDLEIY81HVHNYaEVYMEVCARIQQYELC07W","AF>YVHDLEIY81HVHNYaEVYMEVCARIQQYELC07W"
"GETPPNLINQK0KYDED4IASCQHCPMK9VGPPHAAAIGYGHGEQDRFTGETLNeI","GETPPNLINQK0KYDED4IASCQHCPMK9VGPPHAAAIGYGHGEQDRFTGETLNeI LVIPMK9VGPPHAAAIGYGHGEQDRFTGETLNeI LGIICQSRWWTWRHGPPTAAIGIAEHCGI pHVQCE"
"YP_009481LSLRFQRE","YP_009481LSLRFQRE"
"VDLSE1GSSILQHGLCDIE","VDLSE1GSSILQHGLCDIE"
"GYELI5YQTVPLPTPGHIPTESEs_CNKPRFNHNCDEAGNIFGCNTFK[ECSTGNEAEPHHEPH4AGWPLGYSAERFDVFEng","GYELI5YQTVPLPTPGHIPTESEs_CNKPRFNHNCDEAGNIFGCNTFK[ECSTGNEAEPHHEPH4AGWPLGYSAERFDVFEng LVCYRCIARAADWHPWYAVILTNIGKNCYSSGCRGLV"
"GYELI5YQTVPLPTPGHIPTESEsIGPRcAF>YIPKRNFRPLVIVPCFLIYNQMRIQRKMYHAruDNIaTL","GYELI5YQTVPLPTPGHIPTESEsIGPRcAF>YIPKRNFRPLVIVPCFLIYNQMRIQRKMYHAruDNIaTL SQE"
"ACIteDREKAAF>YYYYYEVCPDNrn","ACIteDREKAAF>YYYYYEVCPDNrn QGAAGRTLAAIGaLSICPFQELLC"
"PHTGNELIPLNTWGRPDANQP","PHTGNELIPLNTWGRPDANQP"
"GEMRGMGGYE","GEMRGMGGYE"
"E4AVRQCRCFIWVPHAMHGTDFPPAHKAEPDTEEIY8MWVQMRCLMLaEDHS","E4AVRQCRCFIWVPHAMHGTDFPPAHKAEPDTEEIY8MWVQMRCLMLaEDHS"
"IKAGVEHTNGEKAGVDDRFPLMGFLHKTQGFDHoLLoraPLRMGPQWPYFFADQPDNVCPQQWVTHoLLSPSGHHVKGFAGACWCMSdrGCHQSQNL>SNVITRSGQMs","IKAGVEHTNGEKAGVDDRFPLMGFLHKTQGFDHoLLoraPLRMGPQWPYFFADQPDNVCPQQWVTHoLLSPSGHHVKGFAGACWCMSdrGCHQSQNL>SNVITRSGQMs GVVWRGAF>YHNLRAVRYA"
"YMVPNPNGSTGIFVVVQVRLMHPRYSIAMGF9HNGQSLGVHVIGGDGVINMEVSEGEAGSELTICVLSTQGIDCEHCGGHSDLNPQSSnNSAGHSNWQDLSMDTPTADGIAFSP","YMVPNPNGSTGIFVVVQVRLMHPRYSIAMGF9HNGQSLGVHVIGGDGVINMEVSEGEAGSELTICVLSTQGIDCEHCGGHSDLNPQSSnNSAGHSNWQDLSMDTPTADGIAFSP"
"YTQLHCASL6LSEYPVHAV","YTQLHCASL6LSEYPVHAV"
"GYELI5ELCQSVDVVHRYQLCCMWLLads_CCCCCGHGQGCNYFFQ","GYELI5ELCQSVDVVHRYQLCCMWLLads_CCCCCGHGQGCNYFFQ"
"GDTWFMHIAQINK[PLAPTSINHVWGPDTLSQEINKLGDGHNINGHSGVVWL","GDTWFMHIAQINK[PLAPTSINHVWGPDTLSQEINKLGDGHNINGHSGVVWL"
"GYELI5ELCQSVDVVHRYQNICQ","GYELI5ELCQSVDVVHRYQNICQ"
"LQILKHWDVEVQREANPKEGCRFGLRjWjGHKVIIMMVQ1TVFAGVoNNHQleTHlHIGVQ9TSPTIWLRNQRKQPCDLPFFGNCDGRAPWRLCPHSDTIMEIIIQNYPVSLETGD>YMGRGHLLGIAPLjWANGNQmQVKEYTELDEHDDYEIRPCAVTE7KTQRHRLCG","LQILKHWDVEVQREANPKEGCRFGLRjWjGHKVIIMMVQ1TVFAGVoNNHQleTHlHIGVQ9TSPTIWLRNQRKQPCDLPFFGNCDGRAPWRLCPHSDTIMEIIIQNYPVSLETGD>YMGRGHLLGIAPLjWANGNQmQVKEYTELDEHDDYEIRPCAVTE7KTQRHRLCG"
"CEFAPLGWMLaEDHLTIHTDRWIFPTKGPTWWQKKIDADNoSDVMGIQRGHTAGPPN>YVWQTGTTPSAIATWQKKHAHVYCGHINSVGL9LDDYTTGDHR","CEFAPLGWMLaEDHLTIHTDRWIFPTKGPTWWQKKIDADNoSDVMGIQRGHTAGPPN>YVWQTGTTPSAIATWQKKHAHVYCGHINSVGL9LDDYTTGDHR"
```

I think it might result from the ID duplication as you suggested. Could you give me some hint about how to fix this?

Thanks.

Dot

**Comment by mw55309** [comment]
Are you able to make the specific file you're trying to parse available?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by shafferm** [comment]
Thanks for the help looking into this @mw55309.

@dot4822, if the viral refseq fasta is still there can you try checking for any duplicate headers? The file would be called something like `viral.%s.protein.faa.gz` with the %s replaced with the date you downloaded it. Here is a python script to do that:
```python
from skbio import read as read_sequence
from collections import counter

YOUR_FASTA_FILE = '/path/to/your/viral_refseq.fasta'

ids = Counter([seq.metadata['id'] for seq in read_sequence(YOUR_FASTA_FILE, format='fasta')])
for header, count in ids.most_common():
    if count > 1:
        print(header, count)

```

Alternately have you tried rerunning the setup if that is feasible?

**Comment by dot4822** [comment]
Hi Mick and Michael,

I re-installed DRAM from Miniconda (to solve the hmmer 3.3.1 issue) and re-ran the setup. Now it works well. So it seems that there was something wrong with the viral database file downloaded previously during DRAM setup.

Thanks for all your helps.

Dot

**Comment by Dorothydu12** [comment]
> Hi Mick and Michael,
> 
> I re-installed DRAM from Miniconda (to solve the hmmer 3.3.1 issue) and re-ran the setup. Now it works well. So it seems that there was something wrong with the viral database file downloaded previously during DRAM setup.
> 
> Thanks for all your helps.
> 
> Dot

**Comment by Dorothydu12** [comment]
> Hi Mick and Michael,
> 
> I re-installed DRAM from Miniconda (to solve the hmmer 3.3.1 issue) and re-ran the setup. Now it works well. So it seems that there was something wrong with the viral database file downloaded previously during DRAM setup.
> 
> Thanks for all your helps.
> 
> Dot

Hi Dot,

I met the same error, did u use the install pipeline "wget https://raw.githubusercontent.com/shafferm/DRAM/master/environment.yaml
conda env create -f environment.yaml -n DRAM"

Thanks,
Dorothy
